File: evita/visuals/dom/document.cc
// Copyright (c) 2015 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "evita/visuals/dom/document.h"

#include "base/logging.h"
#include "evita/visuals/dom/element.h"

namespace visuals {

//////////////////////////////////////////////////////////////////////
//
// Document::LockScope
//
Document::LockScope::LockScope(const Document& document) : document_(document) {
  document_.Lock();
}

Document::LockScope::~LockScope() {
  document_.Unlock();
}

//////////////////////////////////////////////////////////////////////
//
// Document
//
Document::Document() : ContainerNode(this, L"#document") {}
Document::~Document() {}

void Document::AddObserver(DocumentObserver* observer) const {
  const_cast<Document*>(this)->observers_.AddObserver(observer);
}

ElementNode* Document::GetElementById(base::StringPiece16 id_value) const {
  AtomicString id(id_value);
  const auto& it = id_map_.find(id);
  if (it == id_map_.end())
    return nullptr;
  // TODO(eval1749): We should return a first node in tree order.
  return it->second;
}

void Document::Lock() const {
  ++lock_count_;
}

void Document::RegisterElementIdIfNeeded(const ElementNode& element) {
  DCHECK(!is_locked());
  if (element.id().empty())
    return;
  const auto& result = id_map_.insert(
      std::make_pair(element.id(), const_cast<ElementNode*>(&element)));
  DCHECK(result.second) << "id_map_ already has " << element;
}

void Document::RemoveObserver(DocumentObserver* observer) const {
  const_cast<Document*>(this)->observers_.RemoveObserver(observer);
}

void Document::Unlock() const {
  DCHECK_GT(lock_count_, 0);
  --lock_count_;
}

void Document::UnregisterElementIdIfNeeded(const ElementNode& element) {
  DCHECK(!is_locked());
  if (element.id().empty())
    return;
  const auto& start = id_map_.find(element.id());
  DCHECK(start != id_map_.end()) << "id_map_ should have " << element;
  for (auto it = start; it != id_map_.end() && it->first == element.id();
       ++it) {
    if (it->second == &element) {
      id_map_.erase(it);
      return;
    }
  }
  DVLOG(ERROR) << "id_map_ should have " << element;
}

// gc::Visitable
void Document::Accept(gc::Visitor* visitor) {
  for (const auto& entry : id_map_)
    visitor->Visit(entry.second);
}

}  // namespace visuals

File: evita/dom/visuals/document.js
// Copyright (c) 2016 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

goog.require('visuals.Node');

goog.scope(function() {

/** @constructor */
const Element = visuals.Element;

/** @constructor */
const Image = visuals.Image;

/** @constructor */
const Node = visuals.Node;

/** @constructor */
const Text = visuals.Text;

class Document extends Node {
  constructor() { super(null, NodeHandle.createDocument()); }

  /**
   * @param {string} tagName
   * @param {string=} id
   * @return {!Element}
   */
  createElement(tagName, id = '') {
    const handle = NodeHandle.createElement(this.handle_, tagName, id);
    return new Element(this, handle, tagName, id);
  }

  /**
   * @param {!ImageData} data
   * @return {!Image}
   */
  createImage(data) {
    const handle = NodeHandle.createImage(this.handle_, data);
    return new Image(this, handle, data);
  }

  /**
   * @param {!Uint8Array} data
   * @return {!Shape}
   */
  createShape(data) {
    const handle = NodeHandle.createShape(this.handle_, data);
    return new Shape(this, handle, data);
  }

  /**
   * @param {string} data
   * @return {!Text}
   */
  createText(data) {
    const handle = NodeHandle.createText(this.handle_, data);
    return new Text(this, handle, data);
  }

  /** @override @return {string} */
  get nodeName() { return '#document'; }
}

/** @constructor */
visuals.Document = Document;
});

/** @constructor */
var Document = visuals.Document;
